### DISCUSSION TITLE
posts/pawnyable-0302/

### DISCUSSION THREAD

**Discussion by giscus** [brant-ruan/fernweh-comments#33] [Announcements]
# posts/pawnyable-0302/

Linux Kernel PWN | 040302 Pawnyable之双取

https://blog.wohin.me/posts/pawnyable-0302/

**Comment by SiriusHsh** [comment]
请问一下博主，`0xffffffff810bf813: add rsp, 0x140; mov eax,r9d; pop rbx; pop r12; pop r13; pop r14; pop r15; pop rbp; ret;`这条gadget是通过什么方式找到的？我尝试了`ropr`、`Ropper`、`ROPgadget`这三个工具都没法找到这条gadget，但在GDB调试中可以看到确实有这条gadget
```
pwndbg> x/10i 0xffffffffa1c00000+0xbf813
   0xffffffffa1cbf813:  add    rsp,0x140
   0xffffffffa1cbf81a:  mov    eax,r9d
   0xffffffffa1cbf81d:  pop    rbx
   0xffffffffa1cbf81e:  pop    r12
   0xffffffffa1cbf820:  pop    r13
   0xffffffffa1cbf822:  pop    r14
   0xffffffffa1cbf824:  pop    r15
   0xffffffffa1cbf826:  pop    rbp
   0xffffffffa1cbf827:  ret
   0xffffffffa1cbf828:  mov    rax,QWORD PTR [rbp-0x48]
```
非常想知道博主是通过什么方式找到的

**Reply by brant-ruan** [reply]
你好，间隔时间有点久，我也不记得当时怎么找到的了，试了一下ropr和ROPgadget确实没发现，不过也有一些类似的gadgets。你观察的很细致！

**Reply by SiriusHsh** [reply]
主要是其他类似的gadget都凑不到0x170的偏移 &#128514;  最接近的一个也是0x168，所以博主你能找到这个可用的gadget很厉害

**Reply by brant-ruan** [reply]
这么看来，当时应该把找的过程记录下来的😅

**Reply by brant-ruan** [reply]
我又想了一下，因为这个需求其实非常具体，这个问题如果转化为“以add rsp的机器指令开头，以ret结尾，中间有n个pop，其中add rsp指令的数值与n*8相加之和等于0x170”按照这个写一个类似正则的二进制文件内容搜索，是不是可以解决？
